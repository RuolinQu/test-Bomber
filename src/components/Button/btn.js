import React from 'react'

import classnames from 'classnames'

export const Button = (props) => {

    const {className,btnType,disabled,size,href,children,...restProps}=props
    
    const classes = classnames('btn',className,
        {
            [`btn-${btnType}`]: btnType,
            [`btn-${size}`]: size,
            'disabled':(btnType==='link')&&disabled
        }
    )
    if (href && btnType === "link") {
        return (
            <a
                className={classes}
                href={href}
                {...restProps}
            >
                {children}
            </a>
        )
    } else {
        return (
            <button
                className={classes}
                disabled={disabled}
                {...restProps}
            >
                {children}
            </button>
        )
    }
}

export default Button