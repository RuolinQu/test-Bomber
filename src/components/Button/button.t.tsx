// import Enzyme,{ shallow } from 'enzyme';
// import Button from './btn'
// import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

// Enzyme.configure({ adapter: new Adapter() });


// describe('Button', () => {
//     it('renders a button', () => {
//         const wrapper = shallow(<Button>Click</Button>);
//         console.log(wrapper.find(Button).get(0).props);
//     //   expect(wrapper.find('.btn').prop('style')).toHaveProperty('backgroundSize', '100%');
//     });
//   });



// import Button from './button'
// import {render,screen,fireEvent} from '@testing-library/react'
// import renderer from 'react-test-renderer';

// describe('Button component', () => {

//     it('should render correctly', () => {
//         const btn = renderer.create(<Button>Click</Button>).toJSON()
//         expect(btn).toMatchSnapshot();
//     })

//     it('should create a correct default button', () => {
//         render(<Button>Click</Button>)
//         const element = screen.getByText('Click')
//         expect(element).toBeInTheDocument();
//         expect(element).toHaveClass('btn btn-default')
//         expect(element.disabled).toBeFalsy()
//     })

//     it('should run the callback once after one click', () => {
//         const handleCilck=jest.fn()
//         render(<Button onClick={handleCilck}>Click</Button>)
//         const element = screen.getByText('Click')
//         expect(handleCilck).toHaveBeenCalledTimes(0)
//         fireEvent.click(element)
//         expect(handleCilck).toHaveBeenCalledTimes(1)
//     })
// })