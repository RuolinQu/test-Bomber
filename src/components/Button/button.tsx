import React, { FC, ButtonHTMLAttributes, AnchorHTMLAttributes } from 'react'

import classnames from 'classnames'

export type ButtonSize = 'lg' | 'sm'
export type ButtonType = 'primary' | 'default' | 'danger' | 'link'


interface BaseButtonProps {
    className?: string;
    btnType?: ButtonType;
    size?: ButtonSize;
    href?: string;
    disabled?: boolean;
    children?:React.ReactNode;
}

type NativeButtonProps = BaseButtonProps & ButtonHTMLAttributes<HTMLElement>
type AnchorButtonProps = BaseButtonProps & AnchorHTMLAttributes<HTMLElement>
export type ButtonProps = Partial<NativeButtonProps & AnchorButtonProps>

export const Button: React.FC<ButtonProps> = ({ className,
    btnType='default',
    disabled=false,
    size,
    href,
    children,
    ...restProps
}) => {
    
    const classes = classnames('btn',className,
        {
            [`btn-${btnType}`]: btnType,
            [`btn-${size}`]: size,
            'disabled':(btnType==='link')&&disabled
        }
    )
    if (href && btnType === "link") {
        return (
            <a
                className={classes}
                href={href}
                {...restProps}
            >
                {children}
            </a>
        )
    } else {
        return (
            <button
                style={{position:'relative'}}
                className={classes}
                disabled={disabled}
                {...restProps}
            >
                {children}
            </button>
        )
    }
}

export default Button