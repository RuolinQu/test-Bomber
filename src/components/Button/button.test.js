
import Button from './button'
import {render,screen,fireEvent} from '@testing-library/react'
import renderer from 'react-test-renderer';

describe('Button component', () => { 
    it('should create a correct default button', () => {
        render(<Button>Click</Button>)
        const element = screen.getByText('Click')
        expect(element).toBeInTheDocument();
        expect(element).toHaveStyle(`position:relative`)
    })
})


// import Enzyme, { shallow, mount } from 'enzyme';
// // import {render,screen,fireEvent} from '@testing-library/react'
// import BTN from './btn'
// import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
// import ReactDOM from "react-dom";

// Enzyme.configure({ adapter: new Adapter() });

// describe('Button', () => {
//     it('renders a button', () => {
//         const wrapper = shallow(<BTN
//             size="lg"
//             btnType="primary"
//             disabled={false}
//         >
//           me
//         </BTN>);

//         // let containerStyle = wrapper.get(0).style;
//         // console.log(containerStyle)
//         // expect(containerStyle).to.have.property('position', 'relative');
        

//         // const elem = wrapper.find(BTN);
//         // expect(getComputedStyle(elem.getDOMNode()).getPropertyValue('position')).toBe('relative');
        
//         expect(wrapper).toHaveStyle(`position:relative`)
//         // expect(wrapper.find('.btn').prop('style')).toHaveProperty('position', 'relative');
//         // console.log(wrapper.first().getElement().type);
//     //   expect(wrapper.find('.btn').prop('style')).toHaveProperty('backgroundSize', '100%');
//     });
//   });