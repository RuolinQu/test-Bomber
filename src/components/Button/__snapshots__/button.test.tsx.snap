// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render correctly 1`] = `
<button
  className="btn btn-default"
  disabled={false}
>
  Click
</button>
`;
